File: android/src/main/java/ai/founderos/mobiletracker/EventQueue.kt
package ai.founderos.mobiletracker

import ai.founderos.mobiletracker.models.Event
import java.util.LinkedList

/**
 * Thread-safe event queue with FIFO eviction policy
 * 
 * This class manages a queue of events with a maximum size limit.
 * When the queue reaches its maximum size, the oldest events are
 * automatically evicted to make room for new events (FIFO eviction).
 * 
 * All operations are thread-safe using synchronized blocks.
 * 
 * @property maxSize Maximum number of events the queue can hold
 */
class EventQueue(private val maxSize: Int) {
    private val events = LinkedList<Event>()
    
    /**
     * Adds an event to the queue
     * 
     * If the queue is at maximum capacity, the oldest event is removed
     * before adding the new event (FIFO eviction policy).
     * 
     * @param event The event to add to the queue
     */
    @Synchronized
    fun enqueue(event: Event) {
        // If queue is at max size, remove oldest event (FIFO eviction)
        if (events.size >= maxSize) {
            events.removeFirst()
        }
        events.addLast(event)
    }
    
    /**
     * Retrieves and removes all events from the queue
     * 
     * @return List of all events that were in the queue
     */
    @Synchronized
    fun dequeue(): List<Event> {
        val eventList = events.toList()
        events.clear()
        return eventList
    }
    
    /**
     * Removes all events from the queue
     */
    @Synchronized
    fun clear() {
        events.clear()
    }
    
    /**
     * Returns the current number of events in the queue
     * 
     * @return The size of the queue
     */
    @Synchronized
    fun size(): Int {
        return events.size
    }
}

File: ios/MobileTracker/EventQueue.swift
import Foundation

/// Thread-safe event queue with FIFO eviction policy
class EventQueue {
    // MARK: - Properties
    
    /// Maximum number of events the queue can hold
    private let maxSize: Int
    
    /// Internal storage for events
    private var events: [Event] = []
    
    /// Serial dispatch queue for thread-safe operations
    private let queue = DispatchQueue(label: "com.mobiletracker.eventqueue")
    
    // MARK: - Initialization
    
    /// Initialize event queue with maximum size
    /// - Parameter maxSize: Maximum number of events to store (default: 100)
    init(maxSize: Int = 100) {
        self.maxSize = maxSize
    }
    
    // MARK: - Public Methods
    
    /// Add an event to the queue
    /// If queue is at max capacity, removes oldest event (FIFO eviction)
    /// - Parameter event: Event to add to the queue
    func enqueue(_ event: Event) {
        queue.sync {
            // If queue is at max size, remove oldest event (FIFO eviction)
            if events.count >= maxSize {
                events.removeFirst()
            }
            
            // Add new event to the end
            events.append(event)
        }
    }
    
    /// Retrieve and remove all events from the queue
    /// - Returns: Array of all events in the queue
    func dequeue() -> [Event] {
        return queue.sync {
            let allEvents = events
            events.removeAll()
            return allEvents
        }
    }
    
    /// Remove all events from the queue
    func clear() {
        queue.sync {
            events.removeAll()
        }
    }
    
    /// Get current number of events in the queue (for testing/debugging)
    /// - Returns: Number of events currently in the queue
    func count() -> Int {
        return queue.sync {
            return events.count
        }
    }
}
